zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,983 @@
+    /// to the raw_public_inputs column and stores a copy in a
+    /// vector for computing RLC(raw_public_inputs).
+    fn assign_extra_fields(
+        &self,
+        region: &mut Region<'_, F>,
+        extra: ExtraValues,
+        randomness: F,
+        raw_pi_vals: &mut Vec<F>,
+    ) -> Result<[AssignedCell<F, F>; 2], Error> {
+        let mut offset = BLOCK_LEN + 1;
+        // block hash
+        // let block_hash = rlc(extra.block_hash.to_fixed_bytes(), randomness);
+        // region.assign_advice(
+        //     || "block.hash",
+        //     self.raw_public_inputs,
+        //     offset,
+        //     || Ok(block_hash),
+        // )?;
+        // raw_pi_vals[offset] = block_hash;
+        // offset += 1;
+
+        // block state root
+        let state_root = rlc(extra.state_root.to_fixed_bytes(), randomness);
+        let state_root_cell = region.assign_advice(
+            || "state.root",
+            self.raw_public_inputs,
+            offset,
+            || Ok(state_root),
+        )?;
+        raw_pi_vals[offset] = state_root;
+        offset += 1;
+
+        // previous block state root
+        let prev_state_root = rlc(extra.prev_state_root.to_fixed_bytes(), randomness);
+        let prev_state_root_cell = region.assign_advice(
+            || "parent_block.hash",
+            self.raw_public_inputs,
+            offset,
+            || Ok(prev_state_root),
+        )?;
+        raw_pi_vals[offset] = prev_state_root;
+        Ok([state_root_cell, prev_state_root_cell])
+    }
+
+    /// Assign `rpi_rlc_acc` and `rand_rpi` columns
+    #[allow(clippy::type_complexity)]
+    fn assign_rlc_pi(
+        &self,
+        region: &mut Region<'_, F>,
+        rand_rpi: F,
+        raw_pi_vals: Vec<F>,
+    ) -> Result<(AssignedCell<F, F>, AssignedCell<F, F>), Error> {
+        let circuit_len = PiCircuitConfig::<F, MAX_TXS, MAX_CALLDATA>::circuit_len();
+        assert_eq!(circuit_len, raw_pi_vals.len());
+
+        // Last row
+        let offset = circuit_len - 1;
+        let mut rpi_rlc_acc = raw_pi_vals[offset];
+        region.assign_advice(
+            || "rpi_rlc_acc",
+            self.rpi_rlc_acc,
+            offset,
+            || Ok(rpi_rlc_acc),
+        )?;
+        region.assign_advice(|| "rand_rpi", self.rand_rpi, offset, || Ok(rand_rpi))?;
+        self.q_end.enable(region, offset)?;
+
+        // Next rows
+        for offset in (1..circuit_len - 1).rev() {
+            rpi_rlc_acc *= rand_rpi;
+            rpi_rlc_acc += raw_pi_vals[offset];
+            region.assign_advice(
+                || "rpi_rlc_acc",
+                self.rpi_rlc_acc,
+                offset,
+                || Ok(rpi_rlc_acc),
+            )?;
+            region.assign_advice(|| "rand_rpi", self.rand_rpi, offset, || Ok(rand_rpi))?;
+            self.q_not_end.enable(region, offset)?;
+        }
+
+        // First row
+        rpi_rlc_acc *= rand_rpi;
+        rpi_rlc_acc += raw_pi_vals[offset];
```

Comment:
The `offset` here actually has value `circuit_len - 1` (assigned in line 568), but it happens that `raw_pi_vals[offset]` is `0` in all test cases, and `raw_pi_vals[0] = 0`, so it passes.

```suggestion
        rpi_rlc_acc += raw_pi_vals[0];
```